### ISSUE TITLE
Support disambiguation of inherited declarations without refinement

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-spec#1241]
It would be nice (if possible?) to have a special syntax to disambiguate inherited declarations (5.1.1), for brevity and to support tough cases like `C.foo()` below:

```ceylon
interface Base {
    shared formal void foo();
}

interface Left satisfies Base {
    shared actual void foo() {}
}

interface Right satisfies Base {
    shared actual formal void foo();
}

class C() satisfies Left & Right {
    //error: Member foo is inherited ambiguously by C from Left and another subtype of Base and so must be refined by C
    //so:

    shared actual void foo() { (super of Left).foo(); } 
    //but, we can't:
    //error: Member refines a non-default, non-formal member: foo in Left
}
```

**Comment by gavinking** [comment]
I'm actually against this. I very carefully designed the refinement model to explicitly disallow this sort of thing because it results in ambiguity for tooling, etc. It's not explicit that `Left.foo()` is supposed to satisfy the contract of `Right.foo()`, and so if `Left`, `Right`, and `C()` were all in different modules, changes could cause things to break unexpectedly.

**Comment by jvasileff** [comment]
I *think* you may be answering a different question.

My concern isn't the requirement of explicit disambiguation (i.e. for C to assert that `Left.foo` satisfies `Right.foo`'s refinement to `Base.foo`'s contract), but with the interaction between non-`default` non-`formal` methods and this requirement. There doesn't seem to be a way for `C` to pick `Left.foo` as the winner, ironically, because `Left.foo` is so adamant about being picked.

**Comment by gavinking** [comment]
No, that wasn't my point. But now that I look at it from a different angle, I think I can better see _your_ point.

The violated restriction is really that `Left.foo()` isn't declared `default`, and so it can't be refined. But In fact you're _not_ trying to refine `Left.foo()` in any meaningful sense. So the argument is that the semantics of `default` is more heavy-handed than it needs to be.

I guess that's right, and so if we loosened the semantics of `default` we could let this one by. Hrm. Let me sleep on it.

**Comment by gavinking** [comment]
I guess this is probably not for 1.2, by the way, since it would surely impact the backend somewhat.

**Comment by tkaitchuck** [comment]
To clarify, would the proposal be to allow exactly one refinement of foo:
 shared actual void foo() { (super of Left).foo(); }

Or to simply never to require writing nothing at all and simply allow C to
compile?

**Comment by jvasileff** [comment]
@tkaitchuck my thought was the former, possibly with an abbreviated syntax that makes it clear we are not overriding with a different implementation.

I don't have a great idea for a syntax, but perhaps it could even be related to something that is more generally useful for composition/delegation.

**Comment by gavinking** [comment]
> To clarify, would the proposal be to allow exactly one refinement of foo:

Yes, that. We would allow you to refine a non-`formal`, non-`default` method, if the refinement just delegates to the method.
